README.md
# Croupier

![Run Tests](https://github.com/SamJakob/croupier/workflows/Run%20Tests/badge.svg) ![MIT License](https://img.shields.io/github/license/SamJakob/croupier) ![GitHub issues](https://img.shields.io/github/issues/SamJakob/croupier)

A custom SocketCluster client implementation for Dart and Flutter, designed to closely reflect the [SocketCluster JavaScript client](https://github.com/SocketCluster/socketcluster-client)'s API.

Released under an [MIT license][license]. 
Feel free to use or adapt this code as you want, though if you do,
please credit me your README or something.

## Usage
You can install it using the package on [pub.dev](https://pub.dev/packages/croupier) - by adding the latest version to your project's `pubspec.yaml` (instructions on the linked pub.dev page).

A simple usage example:

```dart
import 'package:croupier/croupier.dart';

void main() async {
  
  // Create the client.
  var client = SocketClusterClient(
    hostname: 'localhost',
    port: 3000,
    ackTimeout: 500,
  );
  
  // Connect to the server.
  await client.connect();
  
  // Use the client; (invokes 'myProcedure' on the server.)
  var response = await client.invoke('myProcedure');
  print(response);
  
  // Close the client to clean up resources and
  // cleanly disconnect from the server.
  await client.close();
  
}
```

For more examples and guides, please see the [project wiki][wiki].

## Dart API Documentation
Auto-generated API documentation is available at: https://pub.dev/documentation/croupier/latest/

## Features and bugs
Please file feature requests and bugs at the [issue tracker][tracker].

## Test
Unit tests are included.  
Run `dart pub run test` to execute the unit tests.

[license]: https://github.com/SamJakob/croupier/blob/master/LICENSE
[tracker]: https://github.com/SamJakob/croupier/issues
[wiki]: https://github.com/SamJakob/croupier/wiki


CHANGELOG.md
## 2.0.0
- Migrate to null safety.

## 1.0.1
- Add `CONNECTED` event and `CONNECTION_STATE_CHANGE` event.

## 1.0.0+1
- Minor fixes.

## 1.0.0
- Initial release


test/croupier_test.dart
import 'dart:async';

import 'package:croupier/croupier.dart';
import 'package:test/test.dart';

void main() {
  group('Smoke test', () {
    test('All events have names defined', () {
      var definedNames = [];

      for (var event in SCEvent.values) {
        if (event.name == null) {
          fail('$event does not have a unique event name defined.');
        }

        if (definedNames.contains(event.name)) {
          fail("$event's name has already been defined!");
        }

        definedNames.add(event.name);
      }
    });
  });

  group('Basic tests', () {
    late SocketClusterClient client;

    setUpAll(() {
      client = SocketClusterClient(
        hostname: 'localhost',
        port: 3000,
        ackTimeout: 500,
      );
    });

    test('Client connects explicitly', () {
      expect(client.connect(), completion(isNull));
    });

    test('Client cannot be connected multiple times', () {
      expect(client.connect(), throwsStateError);
    });

    test('Client closes after explicit connect call', () async {
      await client.close();

      expect(client.authState, equals(AuthenticationState.unauthenticated));
      expect(client.state, equals(ConnectionState.closed));
    });

    test('Client connects implicitly', () {
      expect(client.invoke('unknownProc'),
          throwsA(TypeMatcher<TimeoutException>()));
    });

    test('Client reconnects', () async {
      print('Sending shutdown command...');
      client.transmit('shutdown');

      print('Delaying 200ms to allow server shutdown.');
      await Future.delayed(Duration(milliseconds: 200));

      print('Done!');
    });

    test('Client closes after implicit connect call', () async {
      await client.close();

      expect(client.authState, equals(AuthenticationState.unauthenticated));
      expect(client.state, equals(ConnectionState.closed));
    });
  });

  group('Remote Procedure Calls (RPC)', () {
    late SocketClusterClient client;

    setUpAll(() {
      client = SocketClusterClient(
        hostname: 'localhost',
        port: 3000,
        ackTimeout: 500,
      );
    });

    test('Good procedure call', () {
      expect(client.invoke('customProc'), completion(equals('Success')));
    });

    test('Bad procedure call', () async {
      try {
        await client.invoke('customProc', {'bad': true});
        fail("Bad procedure call didn't throw an exception.");
      } catch (ex) {
        expect(ex, TypeMatcher<SocketMessageError>());
        expect((ex as SocketMessageError).name, equals('BadCustomError'));
        expect((ex).message, equals('Server failed to execute the procedure'));
      }
    });

    test('Unknown procedure call', () {
      expect(client.invoke('unknownProc'),
          throwsA(TypeMatcher<TimeoutException>()));
    });

    tearDownAll(() async {
      await client.close();
    });
  });

  group('Local Procedure Calls (Server -> Client)', () {
    late SocketClusterClient client;

    setUpAll(() async {
      client = SocketClusterClient(
        hostname: 'localhost',
        port: 3000,
        ackTimeout: 500,
      );
    });

    test('Local receive', () async {
      var completer = Completer();

      client.registerReceiver(
        'testRecv',
        (data) => {
          completer.complete(data['data']),
        },
      );

      client.transmit('triggerTestRecv');

      expect(completer.future,
          completion(equals('4aa78d81f5ff8f60de71c42c86a80b36')));
    });

    test('Local invoke', () {
      var completer = Completer();

      client.registerReceiver(
        'testInvkResult',
        (data) => {
          completer.complete(data['data']),
        },
      );

      client.registerProcedure('testInvk', (data) {
        print('hello from the other side');
        return 'hello from the other side';
      });

      client.transmit('triggerTestInvk');
      expect(completer.future, completion(equals('hello from the other side')));
    });

    tearDownAll(() async {
      await client.close();
    });
  });

  group('Raw data', () {
    late SocketClusterClient client;

    setUpAll(() async {
      client = SocketClusterClient(
        hostname: 'localhost',
        port: 3000,
        ackTimeout: 500,
      );

      await client.connect();
    });

    test('Raw data', () {
      client.send('hello');
    });

    tearDownAll(() async {
      await client.close();
    });
  });

  group('Housekeeping', () {
    test('Signal end of tests to server...', () async {
      var client = SocketClusterClient(
        hostname: 'localhost',
        port: 3000,
        ackTimeout: 500,
      );

      client.transmit('testFinish');
      await client.close();

      print('All done!');
    });
  });
}


lib/croupier.dart
/// A custom SocketCluster client implementation for Dart and Flutter.
library croupier;

import 'dart:async';
import 'dart:collection';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

part 'src/multiplexed_stream.dart';
part 'src/croupier_reconnect_policy.dart';
part 'src/croupier_impl.dart';

enum ConnectionState {
  closed,
  connecting,
  open,
}

enum AuthenticationState {
  unauthenticated,
  authenticated,
}

enum SCEvent {
  /// Emitted whenever the socket initiates a connection to the server. This
  /// includes reconnects.
  connecting,

  /// Emitted whenever the socket connection to the server is established and
  /// the SocketCluster handshake has completed. This includes reconnects.
  ready,

  /// Emitted whenever the socket disconnects or becomes disconnected from the
  /// server. One can read the [SocketClusterClient.closeCode] and the
  /// [SocketClusterClient.closeReason] from the socket to determine the close
  /// code and reason.
  disconnect,

  /// Emitted when the socket is programmatically closed with a call to
  /// [SocketClusterClient.close]. This is triggered right before the socket
  /// closes unlike [SCEvent.disconnect] which is triggered right after the
  /// close is processed.
  close,

  /// Emitted whenever the authentication state changes. The new authentication
  /// state may be read from [SocketClusterClient.authState].
  authStateChange,

  /// Emitted whenever the connection state changes. The new connection state
  /// may be read from [SocketClusterClient.state].
  connectionStateChange,
}

extension SCEventName on SCEvent {
  String? get name {
    switch (this) {
      case SCEvent.connecting:
        return 'connecting';
      case SCEvent.ready:
        return 'ready';
      case SCEvent.disconnect:
        return 'disconnect';
      case SCEvent.close:
        return 'close';
      case SCEvent.authStateChange:
        return 'authStateChange';
      case SCEvent.connectionStateChange:
        return 'connectionStateChange';
      default:
        return null;
    }
  }
}

abstract class SocketClusterClient {
  /// The hostname of the WebSocket URL to connect to.
  String get hostname;

  /// Whether or not the connection is over HTTP or HTTPS.
  bool get secure;

  /// The port of the WebSocket URL to connect to.
  /// Defaults to 443 if secure or 80 if not.
  int get port;

  /// The path which SocketCluster uses to make the initial handshake for the
  /// WebSocket.
  /// Defaults to '/socketcluster/'
  String get path;

  /// A map of key-value pairs which will be used as the query parameters for
  /// the initial HTTP handshake which will initiate the WebSocket connection.
  Map<String, String>? get query;

  /// Returns the computed URL that the WebSocket client will connect to.
  String get url;

  /// The number of milliseconds to wait whilst attempting a connection to the
  /// server.
  /// Defaults to 20000 (20s).
  late int connectTimeout;

  /// The number of milliseconds to wait for the response to an invoke action.
  /// Defaults to 10000 (10s).
  late int ackTimeout;

  /// The auto-reconnect policy to adhere to if the connection to the socket
  /// server fails. Refer to [ReconnectPolicy] for defaults, etc.
  late ReconnectPolicy reconnectPolicy;

  /// The list of sub-protocols to pass to the WebSocket server on connect.
  List<String>? protocols;

  /// The list of headers to pass to the WebSocket server on connect.
  Map<String, dynamic>? headers;

  /// The id of the socket connection. This is `null` initially and will change
  /// each time a new underlying connection is made.
  String? get id;

  /// The id of the socket client. This does not change between connections.
  String? get clientId;

  /// Changes the authentication token currently associated with the socket.
  /// Set this to `null` to indicate that no token should be associated with
  /// the socket.
  set authToken(String authToken);

  /// The current connection state of this socket.
  ConnectionState get state;

  /// The last known authentication state of the socket.
  AuthenticationState get authState;

  /// The interval in milliseconds between pings, as defined by the server upon
  /// handshaking. This will initially be null, until the handshake is complete
  /// and the value has been received from the server.
  int? get pingInterval;

  /// Gets the close code from the underlying WebSocket which is set when the
  /// WebSocket is closed. If there is no close code available, this property
  /// will be [null].
  int? get closeCode;

  /// Gets the close reason from the underlying WebSocket which is set when the
  /// WebSocket is closed. If there is no close reason available, this property
  /// will be [null].
  String? get closeReason;

  factory SocketClusterClient({
    required String hostname,
    bool secure = false,
    int? port,
    String path = '/socketcluster/',
    Map<String, String>? query,
    int connectTimeout = 20000,
    int ackTimeout = 10000,
    ReconnectPolicy reconnectPolicy = const ReconnectPolicy(),
    String? authToken,
    String? clientId,
  }) =>
      _SocketClusterClientImpl(
        hostname: hostname,
        secure: secure,
        port: port,
        path: path,
        query: query,
        connectTimeout: connectTimeout,
        ackTimeout: ackTimeout,
        reconnectPolicy: reconnectPolicy,
        authToken: authToken,
        clientId: clientId,
      );

  /// Connects the client socket to its origin server.
  Future connect();

  /// Send some raw data to the server. This will trigger a 'raw' event on the
  /// server-side which will carry the provided data.
  void send(String data);

  /// Transmits the specified event to the corresponding receiver on the
  /// server-side socket. Any JSON-compatible auxiliary data may be passed as
  /// [data].
  ///
  /// This method does not return anything, will not throw or reject and cannot
  /// be awaited.
  void transmit(
    String event, [
    dynamic data,
    Options options = const Options(),
  ]);

  /// Invoke the specified [procedure] (RPC) on the corresponding server-side
  /// socket. Any JSON-compatible auxiliary data may passed as [data].
  /// The returned [Future] is the response from the server.
  ///
  /// **NOTE:** If the [options] property has [Options.expectResponse] set to
  /// false, it will be implicitly overridden to true by the use of the [invoke]
  /// method. If you are not expecting a response, you should use [transmit]
  /// instead.
  ///
  /// There is a default timeout of 10 seconds to receive a response from the
  /// server as defined on the client ([ackTimeout]). This limit can be changed
  /// by either changing the [ackTimeout] property in the client constructor OR
  /// by setting the [Options.ackTimeout] on the [options] property for the
  /// invoke call.
  Future<dynamic> invoke(
    String procedure, [
    dynamic data,
    Options options = const Options(expectResponse: true),
  ]);

  /// Registers [callback] as an event handler for [event].
  ///
  /// To cancel the callback, you can call `.cancel` on the returned
  /// [StreamSubscription].
  StreamSubscription on(SCEvent event, Function callback);

  /// Registers [callback] as an event handler for the raw event - which is
  /// triggered whenever the server socket calls `socket.send(...)`. The object
  /// produced will have a 'message' property.
  ///
  /// THIS WILL NOT BE CALLED IF THE DATA CANNOT BE JSON-DECODED. You will need
  /// to use [onMessage] for that.
  ///
  /// To cancel the callback, you can call `.cancel` on the returned
  /// [StreamSubscription].
  StreamSubscription onRaw(Function(String data) callback);

  /// Registers [callback] as an event handler for the message event - which is
  /// triggered whenever a message is emitted through the socket.
  ///
  /// To cancel the callback, you can call `.cancel` on the returned
  /// [StreamSubscription].
  StreamSubscription onMessage(Function(String message) callback);

  /// Registers [callback] as a handler for the server-side transmitted event
  /// of [name]. This is essentially the reverse of the [transmit] function,
  /// i.e. when the server calls `.transmit(...)` on the client, this function's
  /// callback is called.
  StreamSubscription registerReceiver(
    String name,
    Function(dynamic data) callback,
  );

  /// Registers [callback] as a handler for the server-side transmitted remote
  /// procedure call (RPC) of [name]. This is essentially the reverse of the
  /// [invoke] function, i.e. when the server calls `.invoke(...)` on the
  /// client, this function's callback is called.
  StreamSubscription registerProcedure(
    String name,
    Function(dynamic data) callback,
  );

  /// Cleans up local data about the socket and explicitly closes the connection
  /// to the server ensuring that the client does not attempt to automatically
  /// reconnect.
  ///
  /// [code] is `1000` by default indicating normal closure and [reason] is null
  /// by default.
  Future<void> close([int code, String reason]);
}

class Options {
  final bool force;
  final bool cloneData;
  final bool expectResponse;
  final int? ackTimeout;
  final bool noTimeout;

  const Options({
    this.force = false,
    this.cloneData = false,
    this.expectResponse = false,
    this.ackTimeout,
    this.noTimeout = false,
  });

  Options copyWith({
    bool? force,
    bool? cloneData,
    bool? expectResponse,
    int? ackTimeout,
    bool? noTimeout,
  }) =>
      Options(
        force: force ?? this.force,
        cloneData: cloneData ?? this.cloneData,
        expectResponse: expectResponse ?? this.expectResponse,
        ackTimeout: ackTimeout ?? this.ackTimeout,
        noTimeout: noTimeout ?? this.noTimeout,
      );
}

class ConnectTimeoutError extends Error {
  final String message;
  ConnectTimeoutError(this.message);

  @override
  String toString() => 'Connect Timeout Error: $message';
}

class NetworkError extends Error {
  final String message;
  NetworkError(this.message);

  @override
  String toString() => 'Network Error: $message';
}

class SocketMessageError extends Error {
  final String? name;
  final String? message;
  final dynamic data;
  SocketMessageError(
    this.data, {
    this.name,
    this.message,
  });

  @override
  String toString() =>
      'Request Error:${name != null ? ' $name' : ''} ${message ?? data}';
}


lib/src/croupier_impl.dart
part of croupier;

class _SocketClusterClientImpl implements SocketClusterClient {
  // static const SC_PROTOCOL_VERSION = 2;

  //**************************************************************************//
  //**************************************************************************//

  // Data passed from base type.

  @override
  String get hostname => _hostname;
  final String _hostname;

  @override
  bool get secure => _secure;
  final bool _secure;

  @override
  int get port => _port;
  final int _port;

  @override
  String get path => _path;
  final String _path;

  @override
  Map<String, String>? get query => _query;
  final Map<String, String>? _query;

  @override
  String get url => _url;
  final String _url;

  @override
  int connectTimeout;

  @override
  int ackTimeout;

  @override
  ReconnectPolicy reconnectPolicy;

  @override
  List<String>? protocols;

  @override
  Map<String, dynamic>? headers;

  @override
  String? get id => _id;
  String? _id;

  @override
  String? get clientId => _clientId;
  final String? _clientId;

  @override
  ConnectionState get state => __state;
  ConnectionState __state = ConnectionState.closed;
  // Internal setter for connection state, to emit the relevant event.
  set _state(ConnectionState _state) {
    __state = _state;
    _emit(SCEvent.connectionStateChange);
  }

  @override
  AuthenticationState get authState => __authState;
  AuthenticationState __authState = AuthenticationState.unauthenticated;
  // Internal setter for connection state, to emit the relevant event.
  set _authState(AuthenticationState _authState) {
    __authState = _authState;
  }

  @override
  int? get pingInterval => _pingInterval;
  int? _pingInterval;

  String? _authToken;
  @override
  set authToken(String authToken) {
    _authToken = authToken;
  }

  @override
  int? get closeCode => _socket?.closeCode;

  @override
  String? get closeReason => _socket?.closeReason;

  //**************************************************************************//
  //**************************************************************************//

  // Private implementation details.

  /// The current call ID, usually implemented as a counter.
  int _cid = -1;

  /// The number of attempts that have been made to reconnect to the server
  /// since the connection was last lost.
  int _reconnectAttemptsMade = 0;

  /// The interval to wait before next attempting to reconnect based on
  /// the current [reconnectPolicy] and the number of [_reconnectAttemptsMade].
  int getReconnectIntervalFor(int attemptsMade) {
    return ((reconnectPolicy.initialDelay +
                    (reconnectPolicy.randomness * Random().nextDouble()))
                .round() *
            pow(reconnectPolicy.multiplier, attemptsMade))
        .round();
  }

  final Map<int, Completer<dynamic>> _expectedResponses;
  final ListQueue<Map<String, dynamic>> _outboundBuffer;

  //**************************************************************************//
  //**************************************************************************//

  // Interface with WebSocket.

  /// The underlying WebSocket connection.
  WebSocket? _socket;

  _MultiplexedStream? _eventsMultiplex;
  _MultiplexedStream? _receiveMultiplex;
  _MultiplexedStream? _invokeMultiplex;
  //_MultiplexedStream _channelDemux;

  //**************************************************************************//
  //**************************************************************************//

  _SocketClusterClientImpl({
    required String hostname,
    bool secure = false,
    int? port,
    String path = '/socketcluster/',
    Map<String, String>? query,
    this.connectTimeout = 20000,
    this.ackTimeout = 10000,
    this.reconnectPolicy = const ReconnectPolicy(),
    this.protocols,
    this.headers,
    String? authToken,
    String? clientId,
  })  : _hostname = hostname,
        _secure = secure,
        _port = port ?? (secure ? 443 : 80),
        _path = path,
        _query = query,
        _url = Uri(
          scheme: secure ? 'wss' : 'ws',
          host: hostname,
          port: port ?? (secure ? 443 : 80),
          path: path,
          queryParameters: query,
        ).toString(),
        _authToken = authToken,
        _clientId = clientId,
        _expectedResponses = {},
        _outboundBuffer = ListQueue<Map<String, dynamic>>(),
        _eventsMultiplex = _MultiplexedStream(allowGlobalChannel: false),
        _receiveMultiplex = _MultiplexedStream(),
        _invokeMultiplex = _MultiplexedStream();

  @override
  Future connect() async {
    if (__state != ConnectionState.closed) {
      return Future.error(StateError(
        "Socket already open. This error occurs when you call .connect() on a socket that's already open.",
      ));
    }

    _eventsMultiplex ??= _MultiplexedStream(allowGlobalChannel: false);
    _invokeMultiplex ??= _MultiplexedStream();
    _receiveMultiplex ??= _MultiplexedStream();
    //_channelEventDemux = _MultiplexedStream();
    //_channelDataDemux = _MultiplexedStream();
    //_receiverDemux = _MultiplexedStream();

    var _connected = false;
    while (!_connected) {
      _cid = -1;
      _reconnectAttemptsMade = 0;
      _state = ConnectionState.connecting;
      _emit(SCEvent.connecting);

      try {
        _socket = await (WebSocket.connect(
          url,
          protocols: protocols,
          headers: headers,
        ).timeout(Duration(milliseconds: connectTimeout), onTimeout: () {
          _onSocketClose();
          throw ConnectTimeoutError(
              'Connect timed out after ${connectTimeout}ms');
        }));

        _connected = true;
      } catch (_) {
        await _reconnectDelay();
        continue;
      }

      _socket!.listen(
        _onSocketMessage,
        onError: _onSocketError,
        onDone: _onSocketClose,
        cancelOnError: false,
      );

      await _onSocketOpen();
    }
  }

  Future<void> _onSocketOpen() async {
    var status = await invoke(
      '#handshake',
      {'authToken': _authToken},
      Options(force: true),
    );

    _id = status['id'];
    _pingInterval = status['pingTimeout'];
    if (status['isAuthenticated']) {
      _authState = AuthenticationState.authenticated;
    } else {
      _authState = AuthenticationState.unauthenticated;
      _authToken = null;
    }
    _reconnectAttemptsMade = 0;

    _state = ConnectionState.open;
    _emit(SCEvent.ready);
    _flushOutboundBuffer();
  }

  void _onSocketMessage(dynamic message) {
    // Handle message
    _emitEvent('message', message.toString());

    // Handle ping
    if (message == '') {
      send('');
      return;
    }

    // Handle list
    if (message is List) {
      for (var element in message) {
        _onSocketMessage(element);
      }

      return;
    }

    try {
      // Handle packet
      var packet = jsonDecode(message);

      if (packet['event'] != null) {
        if (packet['cid'] != null) {
          _emitInvoke(packet['event'], packet);
        } else {
          _emitReceive(packet['event'], packet);
        }

        return;
      }

      if (packet['rid'] != null) {
        if (_expectedResponses.containsKey(packet['rid'])) {
          if (packet['error'] != null) {
            _expectedResponses[packet['rid']]!.completeError(
              SocketMessageError(
                packet['error'],
                name: packet['error']['name'],
                message: packet['error']['message'],
              ),
            );
          } else {
            _expectedResponses[packet['rid']]!.complete(packet['data']);
          }
        }

        return;
      }

      _emitEvent('raw', message);
    } catch (ex, stacktrace) {
      print('Error handling message: $message');
      print(ex);
      print(stacktrace);
    }
  }

  void _onSocketError(Object error, [StackTrace? stackTrace]) {}

  Future<void> _onSocketClose() async {
    var code = _socket!.closeCode;
    _emit(SCEvent.disconnect);

    if (reconnectPolicy.autoReconnect) {
      if (code == null) return await _reconnectDelay();

      // 1005 - close without status
      // 4000 - server ping timeout
      // 4001 - client pong timeout
      if (code != 1000 && code < 4500) {
        // If there's a ping/pong timeout or the server closed without status
        // simply attempt to reconnect immediately as these may simply occur if
        // the client wakes up after a period of inactivity - in which case
        // we want to restore the connection ASAP.
        if (code == 1005 || code == 4000 || code == 4001) {
          return await _reconnectDelay(0);
        }

        return await _reconnectDelay();
      }
    }
  }

  Future<void> _reconnectDelay([int? initialDelay]) async {
    // 1. Compute the timeout.
    var timeout = initialDelay;
    if (timeout == null) {
      if (_reconnectAttemptsMade > 0) {
        timeout = getReconnectIntervalFor(_reconnectAttemptsMade);
      } else {
        timeout = 0;
      }
    }

    timeout = min(timeout, reconnectPolicy.maxDelay);

    // 2. Attempt to reconnect after awaiting the timeout
    await Future.delayed(Duration(milliseconds: timeout));
    await connect();

    // 3. Increment the number of attempts made to reconnect to the server.
    _reconnectAttemptsMade++;
  }

  void _cleanUp() {
    _authState = AuthenticationState.unauthenticated;

    _outboundBuffer.clear();
    _state = ConnectionState.closed;

    //_channelEventDemux.close();
    //_channelDataDemux.close();
    //_receiverDemux.close();

    _eventsMultiplex?.close();
    _eventsMultiplex = null;

    _receiveMultiplex?.close();
    _receiveMultiplex = null;

    _invokeMultiplex?.close();
    _invokeMultiplex = null;
  }

  //**************************************************************************//
  //**************************************************************************//

  // Protocol API

  @override
  void send(String data) {
    if (state != ConnectionState.open) {
      close(1005);
    } else {
      _socket!.add(data);
    }
  }

  @override
  void transmit(
    String event, [
    dynamic data,
    Options options = const Options(),
  ]) {
    _processOutboundEvent(event, data, options);
  }

  @override
  Future<dynamic> invoke(
    String procedure, [
    dynamic data,
    Options options = const Options(expectResponse: true),
  ]) async {
    return await _processOutboundEvent(
      procedure,
      data,
      options.copyWith(expectResponse: true),
    );
  }

  @override
  StreamSubscription on(SCEvent event, Function? callback) {
    return _eventsMultiplex!.subscribeToChannel(event.name, (event) async {
      try {
        if (callback != null) callback();
      } catch (ex, stacktrace) {
        print('An error occurred whilst processing a $event event.');
        print(ex);
        print(stacktrace);
      }
    });
  }

  @override
  StreamSubscription onRaw(Function(String data)? callback) {
    return _eventsMultiplex!.subscribeToChannel('raw', (event) async {
      try {
        if (callback != null) callback(event.event.toString());
      } catch (ex, stacktrace) {
        print('An error occurred whilst processing a raw event.');
        print(ex);
        print(stacktrace);
      }
    });
  }

  @override
  StreamSubscription onMessage(Function(String data)? callback) {
    return _eventsMultiplex!.subscribeToChannel('message', (event) async {
      try {
        if (callback != null) callback(event.event.toString());
      } catch (ex, stacktrace) {
        print('An error occurred whilst processing a raw event.');
        print(ex);
        print(stacktrace);
      }
    });
  }

  @override
  StreamSubscription registerReceiver(
    String name,
    Function(dynamic data)? callback,
  ) {
    return _receiveMultiplex!.subscribeToChannel(name, (
      dynamic event,
    ) async {
      try {
        if (callback != null) callback(event.event);
      } catch (ex, stacktrace) {
        print('An error occurred whilst processing a raw event.');
        print(ex);
        print(stacktrace);
      }
    });
  }

  @override
  StreamSubscription registerProcedure(
    String name,
    Function(dynamic data)? callback,
  ) {
    return _invokeMultiplex!.subscribeToChannel(name, (event) async {
      int rid = event.event['cid'];

      try {
        if (callback != null) {
          var response = await callback(event.event);

          await _processOutboundEvent(
            null,
            response,
            const Options(),
            rid,
          );
        }
      } catch (ex, stacktrace) {
        print('An error occurred whilst processing a raw event.');
        print(ex);
        print(stacktrace);
      }
    });
  }

  @override
  Future<void> close([int? code, String? reason]) async {
    if (__state != ConnectionState.closed) {
      _emit(SCEvent.close);
      _clearExpectedResponses();

      // Close the socket.
      if (_socket != null) await _socket!.close(code ?? 1000, reason);

      // Allow socket time to close and trigger events before cleaning up.
      await Future.delayed(Duration(milliseconds: 10));

      // Clean up.
      _cleanUp();
    }
  }

  //**************************************************************************//
  //**************************************************************************//

  // Protocol Implementation

  Future<dynamic> _processOutboundEvent(
    String? event,
    dynamic data, [
    Options options = const Options(),
    int? rid,
  ]) async {
    if (state == ConnectionState.closed) {
      await connect();
    }

    data ??= {};

    int? cid;
    if (options.expectResponse) cid = ++_cid;

    var outboundPacket = <String, dynamic>{
      'cid': cid,
      'rid': rid,
      'event': event,
      'data': (data is Map && options.cloneData) ? {...data} : data,
    };

    Future? responseFuture;

    // Add an 'expected response' for this call ID to the expected responses
    // array if the response is, in fact, expected.
    if (options.expectResponse) {
      var outboundEventCompleter = Completer();
      _expectedResponses[cid!] = outboundEventCompleter;
      responseFuture = outboundEventCompleter.future;
      if (!options.noTimeout) {
        responseFuture = responseFuture.timeout(Duration(
          milliseconds: options.ackTimeout ?? ackTimeout,
        ));
      }
    }

    // Add the data to the outbound queue.
    _outboundBuffer.add(outboundPacket);

    // ...and flush the queue if the connection is still good.
    if (state == ConnectionState.open || options.force) {
      _flushOutboundBuffer();
    }

    // This will either return null (in the instance we don't care about
    // the response) or it'll return a future that can be awaited to obtain
    // the response assuming it comes back from the server before the timeout
    // is exceeded.
    return responseFuture;
  }

  void _flushOutboundBuffer() {
    _outboundBuffer.removeWhere((outboundPacket) {
      try {
        _socket!.add(jsonEncode(outboundPacket));
        return true;
      } catch (er) {
        return false;
      }
    });
  }

  void _clearExpectedResponses() async {
    for (var expectedResponseKey in _expectedResponses.keys) {
      var expectedResponse = _expectedResponses[expectedResponseKey]!;
      if (!expectedResponse.isCompleted) {
        expectedResponse.completeError(
          NetworkError(
            'Event #$expectedResponseKey was aborted.',
          ),
        );
      }
    }
  }

  void _emit(SCEvent event) {
    _eventsMultiplex!.addToChannel(event.name, null);
  }

  void _emitEvent(String event, [dynamic data]) {
    _eventsMultiplex!.addToChannel(event, data);
  }

  void _emitReceive(String event, dynamic data) {
    _receiveMultiplex!.addToChannel(event, data);
  }

  void _emitInvoke(String event, dynamic data) {
    _invokeMultiplex!.addToChannel(event, data);
  }
}


testserver/index.js
const http = require('http');
const socketClusterServer = require('socketcluster-server');

let httpServer = http.createServer();
let agServer = socketClusterServer.attach(httpServer);

(async () => {
  // Handle new inbound sockets.
  for await (let {socket} of agServer.listener('connection')) {

    console.log(`Socket ${socket.id} connected.`);

    (async () => {
      for await (let data of socket.receiver('shutdown')) {
        console.log("Shutdown command issued...");

        await new Promise((resolve, reject) => {
          httpServer.close((er) => {
            if (er) return reject();
            resolve();
          });
        });

        console.log("Restarting...");
        await httpServer.listen(3000);
        console.log("Server listening on localhost:3000");
      }
    })();

    (async () => {
      for await (let data of socket.receiver('testFinish')) {
        console.log("Tests done!");

        await new Promise(resolve => setTimeout(resolve, 200));

        if (process.env.CLOSE_ON_TEST_FINISH) {
          console.log("Closing server...");
          await new Promise((resolve, reject) => {
            httpServer.close((er) => {
              if (er) return reject();
              resolve();
            });
          });

          console.log("Exiting...");
          process.exit(0);
        }
      }
    })();

    (async () => {
      // Set up a loop to handle remote transmitted events.
      for await (let data of socket.receiver('triggerTestRecv')) {
        socket.transmit('testRecv', '4aa78d81f5ff8f60de71c42c86a80b36');
      }
    })();

    (async () => {
      // Set up a loop to handle remote transmitted events.
      for await (let data of socket.receiver('triggerTestInvk')) {
        let response = await socket.invoke('testInvk');
        console.log(response);
        socket.transmit('testInvkResult', response);
      }
    })();

    (async () => {
      // Set up a loop to handle and respond to RPCs for a procedure.
      for await (let req of socket.procedure('customProc')) {
        if (req.data.bad) {
          let error = new Error('Server failed to execute the procedure');
          error.name = 'BadCustomError';
          req.error(error);
        } else {
          req.end('Success');
        }
      }
    })();

    (async () => {
      // Set up a loop to handle remote transmitted events.
      for await (let data of socket.listener('raw')) {
        console.log(data);
      }
    })();

  }
})();

(async () => {
  await httpServer.listen(3000);
  console.log("Server listening on localhost:3000");
})();

